The "ConvertDllsToWebCil" task failed:  System.NotImplementedException on s390x

### Description

With a .NET 8 Preview 7 SDK  on s390x (IBM Z), running `dotnet new blazorwasm && dotnet build` fails:


```
/usr/lib64/dotnet/sdk/8.0.100-preview.7.23376.1/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(314,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm_template/blazorwasm_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018: The "ConvertDllsToWebCil" task failed unexpectedly. [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018: System.NotImplementedException: The method or operation is not implemented. [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.WriteStructure[WebcilHeader](Stream s, WebcilHeader structure) [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.WriteHeader(Stream s, WebcilHeader header) [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.WriteConversionTo(Stream outputStream, FileStream inputStream, PEFileInfo peInfo, WCFileInfo wcInfo) [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.ConvertToWebcil() [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.WebAssembly.Build.Tasks.WebcilConverter.ConvertToWebcil() [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.Sdk.WebAssembly.ConvertDllsToWebCil.Execute() [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
```


As does `dotnet new blzaorwasm-empty && dotnet build`:


```
/usr/lib64/dotnet/sdk/8.0.100-preview.7.23376.1/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(314,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018: The "ConvertDllsToWebCil" task failed unexpectedly. [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018: System.NotImplementedException: The method or operation is not implemented. [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.WriteStructure[WebcilHeader](Stream s, WebcilHeader structure) [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.WriteHeader(Stream s, WebcilHeader header) [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.WriteConversionTo(Stream outputStream, FileStream inputStream, PEFileInfo peInfo, WCFileInfo wcInfo) [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.ConvertToWebcil() [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.WebAssembly.Build.Tasks.WebcilConverter.ConvertToWebcil() [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.Sdk.WebAssembly.ConvertDllsToWebCil.Execute() [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
```

### Reproduction Steps

`dotnet new blazorwasm && dotnet build` on IBM Z

### Expected behavior

The `dotnet build` works, or the `blazorwasm` entry doesn't show up as a valid template item.

### Actual behavior

Stack trace.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

It's probably from here, which happens when the architecture is big-endian only: https://github.com/dotnet/runtime/blob/9d08b24d743d0c57203a55c3d0c6dd1bc472e57e/src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs#L217

--- a/src/tasks/Microsoft.NET.WebAssembly.Webcil/Microsoft.NET.WebAssembly.Webcil.csproj
+++ b/src/tasks/Microsoft.NET.WebAssembly.Webcil/Microsoft.NET.WebAssembly.Webcil.csproj
@@ -16,6 +16,7 @@
 
   <ItemGroup>
     <!-- we need to keep the version of System.Reflection.Metadata in sync with dotnet/msbuild and dotnet/sdk -->
+    <PackageReference Include="System.Memory" Version="$(SystemMemoryVersion)" /> 
     <PackageReference Include="System.Reflection.Metadata" Version="$(SystemReflectionMetadataVersion)" />
     <PackageReference Include="System.Collections.Immutable" Version="$(SystemCollectionsImmutableVersion)" />
   </ItemGroup>

--- a/src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs
+++ b/src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs
@@ -175,7 +175,7 @@ public class WebcilConverter
                                 SectionStart: firstWCSection);
     }
 
-    private static void WriteHeader(Stream s, WebcilHeader header)
+    private static void WriteHeader(Stream s, WebcilHeader webcilHeader)
     {
         if (!BitConverter.IsLittleEndian)
         {
@@ -187,7 +187,7 @@ public class WebcilConverter
             webcilHeader.pe_debug_rva = BinaryPrimitives.ReverseEndianness(webcilHeader.pe_debug_rva);
             webcilHeader.pe_debug_size = BinaryPrimitives.ReverseEndianness(webcilHeader.pe_debug_size);
         }
-        WriteStructure(s, header);
+        WriteStructure(s, webcilHeader);
     }
 
     private static void WriteSectionHeaders(Stream s, ImmutableArray<WebcilSectionHeader> sectionsHeaders)

--- a/src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs
+++ b/src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs
@@ -177,6 +177,16 @@ public class WebcilConverter
 
     private static void WriteHeader(Stream s, WebcilHeader header)
     {
+        if (!BitConverter.IsLittleEndian)
+        {
+            webcilHeader.version_major = BinaryPrimitives.ReverseEndianness(webcilHeader.version_major);
+            webcilHeader.version_minor = BinaryPrimitives.ReverseEndianness(webcilHeader.version_minor);
+            webcilHeader.coff_sections = BinaryPrimitives.ReverseEndianness(webcilHeader.coff_sections);
+            webcilHeader.pe_cli_header_rva = BinaryPrimitives.ReverseEndianness(webcilHeader.pe_cli_header_rva);
+            webcilHeader.pe_cli_header_size = BinaryPrimitives.ReverseEndianness(webcilHeader.pe_cli_header_size);
+            webcilHeader.pe_debug_rva = BinaryPrimitives.ReverseEndianness(webcilHeader.pe_debug_rva);
+            webcilHeader.pe_debug_size = BinaryPrimitives.ReverseEndianness(webcilHeader.pe_debug_size);
+        }
         WriteStructure(s, header);
     }
 
@@ -201,21 +211,6 @@ public class WebcilConverter
             );
         }
         WriteStructure(s, sectionHeader);
-    }
-
-    private static void WriteStructure(Stream s, WebcilHeader webcilHeader)
-    {
-        if (!BitConverter.IsLittleEndian)
-        {
-            webcilHeader.version_major = BinaryPrimitives.ReverseEndianness(webcilHeader.version_major);
-            webcilHeader.version_minor = BinaryPrimitives.ReverseEndianness(webcilHeader.version_minor);
-            webcilHeader.coff_sections = BinaryPrimitives.ReverseEndianness(webcilHeader.coff_sections);
-            webcilHeader.pe_cli_header_rva = BinaryPrimitives.ReverseEndianness(webcilHeader.pe_cli_header_rva);
-            webcilHeader.pe_cli_header_size = BinaryPrimitives.ReverseEndianness(webcilHeader.pe_cli_header_size);
-            webcilHeader.pe_debug_rva = BinaryPrimitives.ReverseEndianness(webcilHeader.pe_debug_rva);
-            webcilHeader.pe_debug_size = BinaryPrimitives.ReverseEndianness(webcilHeader.pe_debug_size);
-        }
-        WriteStructure(s, webcilHeader);
     }
 
 #if NETCOREAPP2_1_OR_GREATER

--- a/src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs
+++ b/src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs
@@ -2,6 +2,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+using System.Buffers.Binary;
 using System.IO;
 using System.Collections.Immutable;
 using System.Reflection.PortableExecutable;
@@ -181,9 +182,6 @@ public class WebcilConverter
 
     private static void WriteSectionHeaders(Stream s, ImmutableArray<WebcilSectionHeader> sectionsHeaders)
     {
-        // FIXME: fixup endianness
-        if (!BitConverter.IsLittleEndian)
-            throw new NotImplementedException();
         foreach (var sectionHeader in sectionsHeaders)
         {
             WriteSectionHeader(s, sectionHeader);
@@ -192,16 +190,38 @@ public class WebcilConverter
 
     private static void WriteSectionHeader(Stream s, WebcilSectionHeader sectionHeader)
     {
+        if (!BitConverter.IsLittleEndian)
+        {
+            sectionHeader = new WebcilSectionHeader
+            (
+                virtualSize: BinaryPrimitives.ReverseEndianness(sectionHeader.VirtualSize),
+                virtualAddress: BinaryPrimitives.ReverseEndianness(sectionHeader.VirtualAddress),
+                sizeOfRawData: BinaryPrimitives.ReverseEndianness(sectionHeader.SizeOfRawData),
+                pointerToRawData: BinaryPrimitives.ReverseEndianness(sectionHeader.PointerToRawData)
+            );
+        }
         WriteStructure(s, sectionHeader);
+    }
+
+    private static void WriteStructure(Stream s, WebcilHeader webcilHeader)
+    {
+        if (!BitConverter.IsLittleEndian)
+        {
+            webcilHeader.version_major = BinaryPrimitives.ReverseEndianness(webcilHeader.version_major);
+            webcilHeader.version_minor = BinaryPrimitives.ReverseEndianness(webcilHeader.version_minor);
+            webcilHeader.coff_sections = BinaryPrimitives.ReverseEndianness(webcilHeader.coff_sections);
+            webcilHeader.pe_cli_header_rva = BinaryPrimitives.ReverseEndianness(webcilHeader.pe_cli_header_rva);
+            webcilHeader.pe_cli_header_size = BinaryPrimitives.ReverseEndianness(webcilHeader.pe_cli_header_size);
+            webcilHeader.pe_debug_rva = BinaryPrimitives.ReverseEndianness(webcilHeader.pe_debug_rva);
+            webcilHeader.pe_debug_size = BinaryPrimitives.ReverseEndianness(webcilHeader.pe_debug_size);
+        }
+        WriteStructure(s, webcilHeader);
     }
 
 #if NETCOREAPP2_1_OR_GREATER
     private static void WriteStructure<T>(Stream s, T structure)
         where T : unmanaged
     {
-        // FIXME: fixup endianness
-        if (!BitConverter.IsLittleEndian)
-            throw new NotImplementedException();
         unsafe
         {
             byte* p = (byte*)&structure;
@@ -212,9 +232,6 @@ public class WebcilConverter
     private static void WriteStructure<T>(Stream s, T structure)
         where T : unmanaged
     {
-        // FIXME: fixup endianness
-        if (!BitConverter.IsLittleEndian)
-            throw new NotImplementedException();
         int size = Marshal.SizeOf<T>();
         byte[] buffer = new byte[size];
         IntPtr ptr = IntPtr.Zero;

--- a/src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilReader.cs
+++ b/src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilReader.cs
@@ -6,7 +6,7 @@ using System.Collections.Immutable;
 using System.IO;
 using System.Reflection;
 using System.Runtime.InteropServices;
-
+using System.Buffers.Binary;
 using System.Reflection.Metadata;
 using System.Reflection.PortableExecutable;
 
@@ -63,13 +63,19 @@ public sealed partial class WebcilReader : IDisposable
         {
             return false;
         }
-        if (!BitConverter.IsLittleEndian)
-        {
-            throw new NotImplementedException("TODO: implement big endian support");
-        }
         fixed (byte* p = buffer)
         {
             header = *(WebcilHeader*)p;
+        }
+        if (!BitConverter.IsLittleEndian)
+        {
+            header.version_major = BinaryPrimitives.ReverseEndianness(header.version_major);
+            header.version_minor = BinaryPrimitives.ReverseEndianness(header.version_minor);
+            header.coff_sections = BinaryPrimitives.ReverseEndianness(header.coff_sections);
+            header.pe_cli_header_rva = BinaryPrimitives.ReverseEndianness(header.pe_cli_header_rva);
+            header.pe_cli_header_size = BinaryPrimitives.ReverseEndianness(header.pe_cli_header_size);
+            header.pe_debug_rva = BinaryPrimitives.ReverseEndianness(header.pe_debug_rva);
+            header.pe_debug_rva = BinaryPrimitives.ReverseEndianness(header.pe_debug_size);
         }
         if (header.id[0] != 'W' || header.id[1] != 'b'
             || header.id[2] != 'I' || header.id[3] != 'L'
@@ -346,6 +352,7 @@ public sealed partial class WebcilReader : IDisposable
 
     private unsafe ImmutableArray<WebcilSectionHeader> ReadSections()
     {
+        WebcilSectionHeader secheader;
         var sections = ImmutableArray.CreateBuilder<WebcilSectionHeader>(_header.coff_sections);
         var buffer = new byte[Marshal.SizeOf<WebcilSectionHeader>()];
         _stream.Seek(SectionDirectoryOffset + _webcilInWasmOffset, SeekOrigin.Begin);
@@ -357,8 +364,24 @@ public sealed partial class WebcilReader : IDisposable
             }
             fixed (byte* p = buffer)
             {
-                // FIXME endianness
-                sections.Add(*(WebcilSectionHeader*)p);
+                secheader = (*(WebcilSectionHeader*)p);
+            }
+            if (!BitConverter.IsLittleEndian)
+            {
+                sections.Add
+                (
+                    new WebcilSectionHeader
+                    (
+                        virtualSize: BinaryPrimitives.ReverseEndianness(secheader.VirtualSize),
+                        virtualAddress: BinaryPrimitives.ReverseEndianness(secheader.VirtualAddress),
+                        sizeOfRawData: BinaryPrimitives.ReverseEndianness(secheader.SizeOfRawData),
+                        pointerToRawData: BinaryPrimitives.ReverseEndianness(secheader.PointerToRawData)
+                    )
+                );
+            }
+            else
+            {
+                sections.Add(secheader);
             }
         }
         return sections.MoveToImmutable();

